CPA calculation

quantenschaum opened the pull request:
see #397

quantenschaum commented:
Would be nice if you had a look and merge this.

wellenvogel commented:
trying it out - after short time run into crash:
Uncaught TypeError: Cannot read properties of undefined (reading 'warning')
    at storeKeys (AisInfoPage.jsx:65:37)
    at renderWithHooks (react-dom.development.js:15486:18)
    at updateFunctionComponent (react-dom.development.js:19617:20)
    at beginWork (react-dom.development.js:21640:16)
    at HTMLUnknownElement.callCallback (react-dom.development.js:4164:14)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:4213:16)
    at invokeGuardedCallback (react-dom.development.js:4277:31)
    at beginWork$1 (react-dom.development.js:27490:7)
    at performUnitOfWork (react-dom.development.js:26596:12)
    at workLoopSync (react-dom.development.js:26505:5)

wellenvogel commented:
Next problem:
The dialog for selecting the sort order in the aispage does not work correctly any more.
![Screenshot_2025-03-24_17-14-19](https://github.com/user-attachments/assets/97f7cc2f-d23b-4152-9c20-6264539a1a73)

wellenvogel commented:
Next test.
Why should we show a TCPA in red if the DCPA is far outside any interesting range?
![Screenshot_2025-03-24_17-20-20](https://github.com/user-attachments/assets/5eca9ca4-f127-4019-87c2-b57de6e500e8)

And: The TCPA heavily fluctuates although the DCPA remains the same (courses nearly anti-parallel).

wellenvogel commented:
What I don't understand is the code starting at aislayer.js#507:
You compute a time for drawing arcs from courseVectorTime-age.
But the courseVectorTime is just the length of the course vector (after multiplying with the speed) - not related to the message age at all. The message age just gives different starting points for drawing things.
time could become negative this way...

I think if we go for a computation of things based on message age we should centralize this a bit by computing the estimated position already in aisdata (what is already in your code) and reuse this later one.
Btw: the computation of the AIS age is inconsistent:
(1) aisdata.js (computeAisTarget): only the "age" value from the ais object is used - this is only the age of the ais message on the server side. For most cases this is (currently - see below) ok as the data has just been received. But for recomputations this is not OK (currently they are only used when changing some settings - so no problem).
(2)aislayer.js: additionally the difference between display time and receive time is used.
Any way to achieve your goal (i.e. keep the CPA info valid with the current movement of the boat and the AIS target) you would need to recompute all the CPA stuff every time the boat position (and / or course) has changed.
I did not do this until now to avoid some performance issues if you have a large amount of AIS targets. But maybe this is not really a problem.
Could also be a starting point for moving some of this stuff into a WebWorker or a ServiceWorker. This way the computation could run outside of the UI thread (we always have multiple cores today) - and the computations could run more often.

quantenschaum commented:
> trying it out - after short time run into crash: Uncaught TypeError: Cannot read properties of undefined (reading 'warning') 

Well, then `props.current` is `undefined`. Why? This could have happened before as well. I added a check for that.

quantenschaum commented:
> Next problem: The dialog for selecting the sort order in the aispage does not work correctly any more.

I work for me. What is the problem? - I just added `sortFields`. - I see, the `sortFields` are global now, I adjusted the loop to mark the selected field. The actual selection is working, though.

quantenschaum commented:
> Next test. Why should we show a TCPA in red if the DCPA is far outside any interesting range? 

I highlighted values below threshold red, I found this useful, but this is of course optional. I removed the CSS but left the class, so it can be enabled in the `user.css`.

> And: The TCPA heavily fluctuates although the DCPA remains the same (courses nearly anti-parallel).

Do you have example data for this case?

quantenschaum commented:
> What I don't understand is the code starting at aislayer.js#507

`age` in this context is (should be) the time passed since the timestamp of the AIS target data to now. This timestamp may include the ais seconds field, but this computation should not cafe for this.

As in CPA computation the RMV has to start at the time shifted (DR) position of the target, but since some time has already passed, it has to be shorter (if the length ist kept, it should extended to astern of the target). The display of the RMV, which allows to read of the CPA geometrically/visually, must be consistent with the CPA computation. This way this is achieved.

quantenschaum commented:
> you would need to recompute all the CPA stuff every time the boat position ...

Now CPA is recomputed when the AIS data is updated in the GUI, so about every 5s. Right? In the time the AIS data might have changed if new packets have been received, and the own data has changed for sure. I think updating CPA every 5s seems sufficient, I could be faster, but this seems fine to me.

wellenvogel commented:
> As in CPA computation the RMV has to start at the time shifted (DR) position of the target, but since some time (the age) has already passed, it has to be shorter (if the length ist kept, it should extended to astern of the target). The display of the RMV, which allows to read of the CPA geometrically/visually, must be consistent with the CPA computation. This way this is achieved.

The meaning of courseVectorTime is relatively clear:
It gives the user an idea at which position this object will be located after this time from now.
Even if the vector starts at the shifted position (that's the one we assume to be the AIS target right now)  it's length (assuming the same SOG) will have to be the same.
But maybe this semantics does not really make sense for the RMV - in this case the courseVectorTime would be out of the game completely.

quantenschaum commented:
> TCPA heavily fluctuates although the DCPA remains the same (courses nearly anti-parallel).

I cannot reproduce this.

wellenvogel commented on server/avnav_store.py:
```diff
@@ -212,7 +212,7 @@ def setAisValue(self,mmsi,data,source=None,priority=0,timestamp=None):
           sec=data.get("second",60) # 60=timestamp not available
           if 0<=sec<60: # use timestamp from ais seconds
-            delay = (now%60-sec)%60 # delay of message (up to 59s)
+            delay = (time.time()%60-sec)%60 # delay of message (up to 59s)
```
will not work as you cannot expect the system time to be in sync with the UTC time of the sender.
Especially it will not work when receiving data from SignalK as the point in time when data is received in AvNav is completely different from the point in time when it is received in SignalK.

quantenschaum replied:
But it will not work at all with `now=monotontic()`, because the absolute values of `monotontic()` has no actual meaning, so that does not make any sense. Since it is mod 60, it only needs the seconds to be in sync. And with many other applications it, a correctly set system clock is a requirement. If you cannot satisfy this, you can turn this feature off.

wellenvogel commented:
I now (hopefully) merged most of this PR into the new ais-reorganize branch.
I would be interested in some feedback from your side before merging into master.
I guess there should still be some code cleanup - but first tests are Ok on my side.
Things are reorganized a lot - all computations including the server query are now running in a separate worker - that should avoid UI slowdown on large number of AIS targets. My goal was to do as much computation as possible in the worker - especially all the math.
Now the computations are repeated whenever the boat position is updated or the list of targets is received.
In a second step maybe even the conversion to map or pixel coordinates could be executed there - but this would require more sync.
I guess at the end mapping all the icons and drawing on the canvas will still cost some performance (maybe the next step should be webgl).

quantenschaum commented:
Nice! Thanks! I will have time to look at it after :egg: time.

quantenschaum commented:
In the AIS settings one can select 
- CPA/BRG from estimated position
  - the RMV has full (course vector) length starting at estimated position
- ALL targets below DCPA/TCPA threshold can be marked with the warning flag (red)

Right? - Looks fine to me.
Thank you!

quantenschaum commented:
I observed that there is a target  flagged as warning but another non-flagged target is shown in the AIS widget.
IMHO the flagged target with lowest TCPA should be shown in the AIS widget.

![image](https://github.com/user-attachments/assets/7ce409e8-c4f3-4e09-a107-b0b0d8d63931)


![image](https://github.com/user-attachments/assets/8b10773c-4c4c-45ac-a374-e21bd185e0a8)

quantenschaum commented:
Probably yes. - What about the internal clock synced to GPS, independent of the system time?

quantenschaum commented:
for the motion vectors

- The TMV starts at the last received position with length of course vector time * speed.
- The icon at the estimated position is somewhere on the TMV, but has no own TMV
- The RMV now (optionally) starts at the estimated position icon, but has full course vector length.

Isn't this somewhat inconsistent? - In my in Implementation the RMV at the estimated position icon was shortened to the remaining time or shifted such that the estimated position icon was somewhere in the middle on this line. The estimated position icon was moving along the RMV the same way the estimated position icon is moving along the TMV.

wellenvogel commented:
For the motion vectors:
Not exactly (doc most probably needs some update).
In General:
The estimated position ("shadow") is the one we expect the AIS target to right now. Consequently motion vectors (true/relative) start from there.
The endpoint of the TMV is the position we expect the target to be at now+courseVectorTime. For RMV it's not that obvious (from my pov) - but the end should also somehow denote now+courseVectorTime.
The line between received pos and estimated pos is an "estimation vector" - i.e. the course the the AIS target has moved from received time to now. It could be drawn with some different color/size/pattern - but I don't think that this would give more clarity for normal users.
I think if we compute data based on the estimated position it does not make sense to draw TMV and RMV starting at the received position at all.

For the "second" parameter:
I felt that using it could create a new class of problems - so the question is whether it is worse the effort.
If you remember our discussions about the queuing of data it could happen that items are sitting in a receive queue for some time (maxAge).
So we could run into a situation where our received GPS data is outdated for some seconds. This is no problem for all normal use cases - you simple get a time that is slightly off. If we use this to compute the AIS send time we could get wrong results (assume e.g. 01:02:00 is received 2 seconds late and we receive an AIS message with seconds 01 before we receive the GPS message. We would compute the send time to 01:01:01 - wrong by 60s).
If we do not consider seconds we still receive some outdated messages - but they are wrong at most by maxAge + the AIS "air time".
From my pov the design of the AIS messages is rather poor at this point. Just to save some bits the create a lot of trouble...

quantenschaum commented:
OK! But estimated position is always computed without considering the rate of turn.

seconds: I think, it is doable, this is what this parameter is meant for, but for a <60s error correction it is probably not worth the effort.

wellenvogel commented:
release with [20250723](https://www.wellenvogel.de/software/avnav/downloads/release/20250723/?lang=en)
